File: server/heartbeat/sources/jobs-source.js
// Heartbeat source #1 (day one): background Job status. Leaf module — only
// jobs/job-store.js and heartbeat/outbox-store.js (both leaves), plus
// schedule-store.js for its own prior checkState.
//
// This does NOT duplicate Jobs' own completion/failure notifications —
// those already fire directly via addNotification() in worker.js/
// orchestrator.js the instant they happen. What this closes is the real gap
// the whole build exists for: a Tier 1 job outbox row (a confirm-gated
// action escalated, or "OK to start controlling your computer?") today sits
// completely silent — invisible to the user — until they happen to start a
// new conversation themselves, at which point prompt.js's jobsSection()
// finally surfaces it. This source turns "sitting silently in the outbox"
// into a real Heartbeat finding, eligible for the SAME urgency-decision step
// (server/heartbeat/decision.js) every other finding goes through — which is
// what can turn it into genuine proactive speech when the user is around.
//
// **Own-checkState dedup, found necessary by live testing, not assumed up
// front.** engine.js's routeFinding() dedups against an UNDELIVERED outbox
// row — real for a Tier 1/2 verdict, but a Tier 3 verdict never creates one
// at all, so a job that keeps sitting in the exact same "still waiting"
// state produced a fresh finding, a fresh (spent) decision.js model call,
// AND a fresh notification on every single tick forever — confirmed live,
// not hypothetical: the same unresolved job re-notified every ~3 minutes
// with no end. `check()` now remembers the specific underlying outbox
// entry's own id it last reported (via schedule-store.js's checkState, the
// same mechanism commitments-source.js already uses) and only returns a
// Finding again once that id actually changes — a genuinely NEW ask, not
// the same still-unanswered one.

import * as jobStore from '../../jobs/job-store.js';
import { getForJob } from '../outbox-store.js';
import { getItem as getScheduleItem } from '../schedule-store.js';

export const SOURCE_ID = 'jobs';
const CHECK_INTERVAL_MS = 3 * 60 * 1000; // cheap DB read only — no model call itself, safe to check often

/**
 * One item per job currently holding an undelivered Tier 1 outbox row.
 * itemKey is the JOB id (not the outbox row id) — a job can only ever hold
 * one live Tier 1 ask at a time in practice, and keying on the job keeps a
 * fresh ask for the same job from spawning a second, parallel schedule row.
 */
export async function listItems() {
  const active = jobStore.listJobs({ status: ['queued', 'planning', 'running', 'awaiting_decision', 'orphaned'] });
  const items = [];
  for (const job of active) {
    const pendingTier1 = getForJob(job.id).some((e) => e.tier === 1 && !e.deliveredAt);
    if (pendingTier1) items.push({ itemKey: job.id, intervalMs: CHECK_INTERVAL_MS });
  }
  return items;
}

/**
 * Re-reads the job's own outbox — if it's since been delivered (the user
 * already started a turn and saw it, or resolved it directly) there's
 * nothing left to surface. Otherwise returns a Finding, but ONLY the first
 * time this exact outbox entry is seen — see this file's own header
 * comment on why that check is needed at all.
 */
export async function check(itemKey) {
  const jobId = itemKey;
  const job = jobStore.getJob(jobId);
  if (!job) return { finding: null };
  const pending = getForJob(jobId).filter((e) => e.tier === 1 && !e.deliveredAt);
  if (!pending.length) return { finding: null, checkState: null }; // nothing outstanding any more — clear any stale memory of a past ask
  const entry = pending[0];

  const prior = getScheduleItem(SOURCE_ID, itemKey)?.checkState;
  if (prior?.lastReportedOutboxId === entry.id) {
    return { finding: null }; // the exact same still-unanswered ask — already reported, don't re-spend a decision call on it
  }

  return {
    finding: {
      summary: `Background work waiting on you: "${job.title}" — ${entry.summary}`,
      detail: JSON.stringify({ jobId, outboxId: entry.id }),
    },
    checkState: { lastReportedOutboxId: entry.id },
  };
}

export const source = { id: SOURCE_ID, defaultIntervalMs: CHECK_INTERVAL_MS, listItems, check };

File: backend/jarvis/heartbeat/sources/jobs_source.py
"""Background work that is waiting on the user.

Not job completions — those already have their own notification the moment they
happen. What this closes is the real gap: a job parked needing permission sits
completely silent until the user happens to start a conversation, which could be
hours after the thing it needed permission for stopped being useful.

**Its own dedup memory, and this is where it was learned.** A tier-3 verdict
creates no outbox row, so the broker's own duplicate check cannot see it, and a
job sitting in the same "still waiting" state produced a fresh finding, a fresh
model call and a fresh notification on every tick — forever, for one unanswered
question. The fix is to remember the specific outbox row last reported and stay
quiet until that actually changes.
"""

from __future__ import annotations

from typing import Any

from .registry import Source

SOURCE_ID = "jobs"
#: A cheap database read, so it can be frequent.
CHECK_INTERVAL_MS = 3 * 60 * 1000

WAITING_STATUSES = ("queued", "planning", "running", "awaiting_decision", "orphaned")


def list_items() -> list[dict[str, Any]]:
    """One item per job holding an undelivered tier-1 row.

    Keyed on the JOB, not the outbox row: a job holds one live ask at a time, and
    keying on the row would spawn a second schedule entry for each new ask.
    """
    from ...heartbeat import outbox
    from ...jobs import job_store

    items = []
    for job in job_store.list_active_jobs():
        if job.get("status") not in WAITING_STATUSES:
            continue
        waiting = [e for e in outbox.for_job(job["id"])
                   if e["tier"] == 1 and not e["deliveredAt"]]
        if waiting:
            items.append({"itemKey": job["id"], "intervalMs": CHECK_INTERVAL_MS})
    return items


def check(item_key: str) -> dict[str, Any]:
    from ...heartbeat import outbox, schedule_store
    from ...jobs import job_store

    job = job_store.get_job(item_key)
    if job is None:
        return {"finding": None}

    waiting = [e for e in outbox.for_job(item_key) if e["tier"] == 1 and not e["deliveredAt"]]
    if not waiting:
        # Nothing outstanding any more: forget which row was reported, so a
        # genuinely new ask is reported rather than suppressed.
        return {"finding": None, "checkState": None}

    entry = waiting[0]
    item = schedule_store.get_item(SOURCE_ID, item_key)
    reported = ((item or {}).get("checkState") or {}).get("lastReportedOutboxId")
    if reported == entry["id"]:
        return {"finding": None}         # the same unanswered ask

    return {
        "finding": {"summary": f'Background work is waiting on you: "{job["title"]}" — '
                               f'{entry["summary"]}',
                    "detail": {"jobId": item_key, "outboxId": entry["id"]}},
        "checkState": {"lastReportedOutboxId": entry["id"]},
    }


source = Source(id=SOURCE_ID, default_interval_ms=CHECK_INTERVAL_MS,
                list_items=list_items, check=check)
